src/System.Text.Primitives/System/Text/InvariantParser_byte.cs

```diff
@@ -77,40 +141,109 @@ unsafe public static bool TryParse(byte* utf8Text, int index, int length, out by
                     {
-                        value = default(byte);
+                        if (bytesConsumed == 0)
+                        {
+                            value = default(byte);
+                            return false;
+                        }
+                        else
+                        {
+                            return true;
+                        }
+                    }
+                    else if (value > Byte.MaxValue / 10)
+                    {
+                        value = 0;
+                        bytesConsumed = 0;
                         return false;
                     }
-                    else
+                    else if (value > 0 && Byte.MaxValue - value * 10 < nextByteVal) // overflow
                     {
-                        return true;
+                        value = 0;
+                        bytesConsumed = 0;
+                        return false;
                     }
+
+                    byte candidate = (byte)(value * 10 + nextByteVal); // left shift the value and add the nextByte.
+
+                    value = candidate;
+                    bytesConsumed++;
                 }
-                else if (value > Byte.MaxValue / 10)
-                {
-                    value = 0;
-                    bytesConsumed = 0;
-                    return false;
-                }
-                else if (value > 0 && Byte.MaxValue - value * 10 < nextByteVal) // overflow
-                {
-                    value = 0;
-                    bytesConsumed = 0;
-                    return false;
-                }
+            }
+            else if (cultureAndEncodingInfo.IsInvariantUtf16)
+            {
+                bool previousByteNonZero = false;
 
-                byte candidate = (byte)(value * 10 + nextByteVal); // left shift the value and add the nextByte.
+                for (int byteIndex = index; byteIndex < length; byteIndex++) // loop through the byte array
+                {
+                    byte nextByte = utf8Text[byteIndex];
+                    if (nextByte == 0 && previousByteNonZero)
+                    {
+                        previousByteNonZero = false;
+                        bytesConsumed++;
+                    }
+                    else if (previousByteNonZero) // this means we're not looking at an ASCII digit
+                    {
+                        if (bytesConsumed == 0) // check to see if we've processed any digits at all
+                        {
+                            value = default(byte); // if we haven't, set value to 0 and return false
+                            return false;
+                        }
+                        else
+                        {
+                            return true; // otherwise return true
+                        }
+                    }
+                    else
+                    {
+                        byte nextByteVal = (byte)(nextByte - '0');
```

Comment:
>byte nextByteVal = (byte)(nextByte - '0'); [](start = 24, length = 42)

I believe there is something wrong in the logic when encountering characters less than the ordinal value of '0' something like 2F